File: agrotech-back/src/main/java/br/com/agrotech/agrotech_api/service/AlertaService.java
package br.com.agrotech.agrotech_api.service;

import br.com.agrotech.agrotech_api.model.Alerta;
import br.com.agrotech.agrotech_api.repository.AlertaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Classe de serviço responsável pelas regras e disparos de alertas do sistema.
 */
@Service
public class AlertaService {

    @Autowired
    private AlertaRepository alertaRepository;

    /**
     * Busca todas as ocorrências de alertas gerados no sistema.
     *
     * @return Uma lista contendo todas as instâncias registradas de {@link Alerta}.
     */
    public List<Alerta> buscarTodos() {
        return alertaRepository.findAll();
    }

    /**
     * Recupera os detalhes de um evento ou anomalia através de seu identificador.
     *
     * @param id Identificador único do alerta.
     * @return Um {@link Optional} com o {@link Alerta} consultado.
     */
    public Optional<Alerta> buscarPorId(Long id) {
        return alertaRepository.findById(id);
    }

    /**
     * Registra uma nova ocorrência ou anomalia detectada pelo ecossistema.
     *
     * @param alerta Dados do evento que gerou o disparo de severidade.
     * @return O {@link Alerta} gravado no banco de dados.
     */
    public Alerta salvar(Alerta alerta) {
        return alertaRepository.save(alerta);
    }

    /**
     * Atualiza o status, severidade ou descrição de um alerta cadastrado (ex: marcar como resolvido).
     *
     * @param id Identificador do alerta a ser atualizado.
     * @param alertaAtualizado Dados atualizados da anomalia.
     * @return O {@link Alerta} atualizado e mantido no banco de dados.
     * @throws RuntimeException Caso a anomalia informada não exista no banco.
     */
    public Alerta atualizar(Long id, Alerta alertaAtualizado) {
        return alertaRepository.findById(id)
                .map(alerta -> {
                    alerta.setDescricao(alertaAtualizado.getDescricao());
                    alerta.setSeveridade(alertaAtualizado.getSeveridade());
                    alerta.setDataHoraDisparo(alertaAtualizado.getDataHoraDisparo());
                    alerta.setResolvido(alertaAtualizado.getResolvido());
                    if (alertaAtualizado.getDispositivo() != null) {
                        alerta.setDispositivo(alertaAtualizado.getDispositivo());
                    }
                    return alertaRepository.save(alerta);
                })
                .orElseThrow(() -> new RuntimeException("Alerta não encontrado com o ID: " + id));
    }

    /**
     * Remove a ocorrência de um alerta do histórico do sistema.
     *
     * @param id Identificador do alerta a ser removido.
     * @throws RuntimeException Caso a anomalia informada não exista no banco.
     */
    public void deletar(Long id) {
        if (!alertaRepository.existsById(id)) {
            throw new RuntimeException("Alerta não encontrado com o ID: " + id);
        }
        alertaRepository.deleteById(id);
    }
}

File: agrotech-front/src/services/alertaService.ts
import { api } from './api';
import type { Alerta } from '../types/Alerta';

/**
 * Serviço encarregado do consumo dos endpoints referentes aos alertas e anomalias do sistema.
 */
export const alertaService = {
    /**
     * Retorna todas as ocorrências de alertas disparados no ecossistema.
     *
     * @returns {Promise<Alerta[]>} Uma promessa contendo a lista de alertas.
     */
    listarTodos: async (): Promise<Alerta[]> => {
        const response = await api.get<Alerta[]>('/alertas');
        return response.data;
    },

    /**
     * Obtém os detalhes de um evento de alerta pelo seu identificador.
     *
     * @param {number} id - Identificador único do alerta.
     * @returns {Promise<Alerta>} Uma promessa com os dados do alerta.
     */
    buscarPorId: async (id: number): Promise<Alerta> => {
        const response = await api.get<Alerta>(`/alertas/${id}`);
        return response.data;
    },

    /**
     * Registra o disparo de uma nova ocorrência ou anomalia de sensor.
     *
     * @param {Alerta} alerta - Dados de severidade, descrição e equipamento associado.
     * @returns {Promise<Alerta>} Uma promessa com o alerta cadastrado.
     */
    disparar: async (alerta: Alerta): Promise<Alerta> => {
        const response = await api.post<Alerta>('/alertas', alerta);
        return response.data;
    },

    /**
     * Atualiza as informações de um alerta (ex: alterar status para resolvido).
     *
     * @param {number} id - Identificador do alerta.
     * @param {Alerta} alerta - Objeto com o status e informações alteradas.
     * @returns {Promise<Alerta>} Uma promessa contendo o alerta atualizado.
     */
    atualizar: async (id: number, alerta: Alerta): Promise<Alerta> => {
        const response = await api.put<Alerta>(`/alertas/${id}`, alerta);
        return response.data;
    },

    /**
     * Exclui um evento de alerta do histórico do sistema.
     *
     * @param {number} id - Identificador único do alerta a ser removido.
     * @returns {Promise<void>} Promessa resolvida após a deleção.
     */
    deletar: async (id: number): Promise<void> => {
        await api.delete(`/alertas/${id}`);
    },
};